### PR TITLE
CoreFoundation: remove private headers from public directory

### PR DESCRIPTION
**Pull request by compnerd**
Clean up the long standing issue with private headers being pushed into
the public headers directory.  Add a private modulemap so that we can
still access the private headers.

### CODE REVIEW

**Comment by compnerd** [comment]
@swift-ci please test

**Comment by millenomi** [comment]
It nope'd.

**Comment by compnerd** [comment]
@swift-ci please test

**Comment by compnerd** [comment]
@swift-ci please test

**Comment by compnerd** [comment]
Okay, there are some hacks here but, this might be better than what we had previously.

- I desugared some types on Windows as the entire WinSDK is not fully modularized yet
- I had to change some `bool` to `Boolean` (could just include `<stdbool.h>`)
- I had to include `<synchapi.h>` explicitly to deal with the partial modular headers :-(
- `-Xcc -DDEPLOYMENT_RUNTIME_SWIFT` needs to be propagated for `ForFoundationOnly.h` to not include `<objc/message.h>` (could use `__has_include`)

Otherwise, we should no longer be exposing any of the private headers from the import.

**Comment by compnerd** [comment]
@swift-ci please test

**Comment by compnerd** [comment]
@swift-ci please test Linux platform

**Comment by compnerd** [comment]
@swift-ci please test Linux platform

**Comment by compnerd** [comment]
@swift-ci please test Linux platform

**Comment by compnerd** [comment]
@swift-ci please test Linux platform

**Comment by compnerd** [comment]
@swift-ci please test

**Comment by compnerd** [comment]
@swift-ci please test Linux platform

**Comment by compnerd** [comment]
@swift-ci please test Linux platform

**Comment by compnerd** [comment]
@swift-ci please test Linux platform

**Comment by compnerd** [comment]
@swift-ci please test Linux platform

**Comment by compnerd** [comment]
@swift-ci please test Linux platform

**Comment by compnerd** [comment]
@swift-ci please test Linux platform

**Comment by compnerd** [comment]
@swift-ci please test Linux platform

**Comment by compnerd** [comment]
@swift-ci please test Linux platform

**Comment by millenomi** [comment]
@swift-ci please test Linux platform

**Comment by compnerd** [comment]
@swift-ci please test Linux platform

**Comment by compnerd** [comment]
@swift-ci please test Linux platform

**Comment by compnerd** [comment]
@swift-ci please test Linux platform

**Comment by compnerd** [comment]
@swift-ci please test Linux platform

**Comment by compnerd** [comment]
@swift-ci please test Linux platform

**Comment by compnerd** [comment]
@swift-ci please test

**Comment by compnerd** [comment]
@swift-ci please test

**Comment by compnerd** [comment]
@swift-ci please test

**Comment by compnerd** [comment]
@swift-ci please test

**Comment by compnerd** [comment]
@swift-ci please test Linux platform

**Comment by compnerd** [comment]
@swift-ci please test Linux platform

**Comment by compnerd** [comment]
@swift-ci please test Linux platform

**Comment by compnerd** [comment]
@swift-ci please test Linux platform

**Comment by compnerd** [comment]
@swift-ci please test Linux platform

**Comment by compnerd** [comment]
@swift-ci please test Linux platform

File: Sources/Foundation/Data.swift [hunk 1]
```diff
@@ -2083,7 +2083,6 @@ public struct Data : ReferenceConvertible, Equatable, Hashable, RandomAccessColl
     }
 
     // slightly faster paths for common sequences
-    @inlinable // This is @inlinable as an important generic funnel point, despite being a non-trivial initializer.
```
Comment (millenomi) [inline Sources/Foundation/Data.swift]: Why remove inlineability here?

Comment (compnerd) [reply Sources/Foundation/Data.swift]: It was a cascading effect due to the use of `_withStackOrHeapBuffer` getting inlined, making this `inlinable` would expose the `@_implementationOnly` interface :-(.

**Comment by spevans** [comment]
This fails to build under Xcode
```
/Applications/Xcode-beta.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.15.sdk/System/Library/Frameworks/CoreFoundation.framework/Modules/module.modulemap:1:18: error: redefinition of module 'CoreFoundation'
framework module CoreFoundation [extern_c] [system] {
                 ^
/Users/spse/src/swift/swift-corelibs-foundation/DerivedData/Foundation/Build/Products/Debug/usr/local/include/CoreFoundation/module.map:1:8: note: previously defined here
module CoreFoundation [extern_c] [system] {
       ^
/Users/spse/src/swift/swift-corelibs-foundation/Sources/Foundation/NSArray.swift:13:8: error: no such module 'CoreFoundation_Private'
import CoreFoundation_Private
```

**Comment by compnerd** [comment]
Please test with following PRs:
apple/llvm-project#1057

@swift-ci please test Linux platform

**Comment by shahmishal** [comment]
The Swift project moved the default branch to `main` and deleted `master` branch, so GitHub automatically closed the PR. Please re-create pull request with main branch.

More detail about the branch update - https://forums.swift.org/t/updating-branch-names/40412

**Pull request closed without merging**